src/core/middleware/drag.ts

```diff
@@ -0,0 +1,240 @@
+function createPositionMatrix(): PositionMatrix {
+	return {
+		client: { x: 0, y: 0 },
+		offset: { x: 0, y: 0 },
+		page: { x: 0, y: 0 },
+		screen: { x: 0, y: 0 }
+	};
+}
+
+function getPositionMatrix(event: PointerEvent): PositionMatrix {
+	return {
+		client: {
+			x: event.clientX,
+			y: event.clientY
+		},
+		offset: {
+			x: event.offsetX,
+			y: event.offsetY
+		},
+		page: {
+			x: event.pageX,
+			y: event.pageY
+		},
+		screen: {
+			x: event.screenX,
+			y: event.screenY
+		}
+	};
+}
+
+function initNode(node: HTMLElement): void {
+	node.style.touchAction = 'none';
+	node.setAttribute('touch-action', 'none');
+}
+
+const emptyResults = Object.freeze({
+	delta: Object.freeze(createPosition()),
+	isDragging: false
+});
+
+const factory = create({ destroy, icache, invalidator, node });
+
+export const drag = factory(({ middleware: { destroy, icache, invalidator, node } }) => {
+	const nodeMap = icache.getOrSet('nodeMap', () => new WeakMap<HTMLElement, NodeData>());
+
+	function getData(target: HTMLElement): { state: NodeData; target: HTMLElement } | undefined {
+		if (nodeMap.has(target)) {
+			return { state: nodeMap.get(target)!, target };
+		}
+		if (target.parentElement) {
+			return getData(target.parentElement);
+		}
+	}
+
+	function onDragStart(event: PointerEvent) {
+		const dragging = icache.get<HTMLElement>('dragging');
+		if (!event.isPrimary && dragging) {
+			const state = nodeMap.get(dragging)!;
+			state.dragResults.isDragging = false;
+			state.invalidate();
+			icache.set('dragging', undefined);
+			return;
+		}
+		if (event.button !== 0) {
+			return;
+		}
+		const data = getData(event.target as HTMLElement);
+		if (data) {
+			const { state, target } = data;
+			icache.set('dragging', target);
+			state.last = state.start = getPositionMatrix(event);
+			state.dragResults.delta = createPosition();
+			state.dragResults.start = deepAssign({}, state.start);
+			state.dragResults.isDragging = true;
+			state.invalidate();
+
+			event.preventDefault();
+			event.stopPropagation();
+		}
+	}
+
+	function onDrag(event: PointerEvent) {
+		const dragging = icache.get<HTMLElement>('dragging');
+		if (!dragging) {
+			return;
+		}
+		// state cannot be unset, using ! operator
+		const state = nodeMap.get(dragging)!;
+		state.last = getPositionMatrix(event);
+		state.dragResults.delta = getDelta(state.start, state.last);
+		state.invalidate();
+
+		event.preventDefault();
+		event.stopPropagation();
+	}
+
+	function onDragStop(event: PointerEvent) {
+		const dragging = icache.get<HTMLElement>('dragging');
+		if (!dragging) {
+			return;
+		}
+		// state cannot be unset, using ! operator
+		const state = nodeMap.get(dragging)!;
+		state.last = getPositionMatrix(event);
+		state.dragResults.delta = getDelta(state.start, state.last);
+		state.dragResults.isDragging = false;
+		state.invalidate();
+		icache.set('dragging', undefined);
+
+		event.preventDefault();
+		event.stopPropagation();
+	}
+
+	if (!icache.get('initialized')) {
```

Comment:
As `initialized` is always `undefined`, i'm not sure what this check does?

Reply:
Also this differs from the existing `Drag` meta, it will add a new handler for every widget that uses the `drag` middleware. In the a`Drag` meta this was done by having a singleton controller.

Reply:
Good catch; it looks like we can either kill this `initialized` check altogether, always creating new handlers, or we can move the listeners to the module scope.

Reply:
I ended up removing this initialization check, creating listeners for each middleware use, with tests in-place to verify the listener cleanup